Fix bug in TopologyManager hint generation after kubelet restart

klueska opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Previously, the HintProviders for the CPUManager and devicemanager would
attempt to generate (new) hints for already running containers after a
kubelet restart.

This patch adds logic to both the CPUManager and the devicemanager to regenerate hints for containers that already have CPUs and/or devices allocated to them.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #84479 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

klueska committed:
Abstract removeStaleState from reconcileState in CPUManager

This will become especially important as we move to a model where
exclusive CPUs are assigned at pod admission time rather than at pod
creation time.

Having this function will allow us to do garbage collection on these
CPUs anytime we are about to allocate CPUs to a new set of containers,
in addition to reclaiming state periodically in the reconcileState()
loop.

klueska commented:
/assign @ConnorDoyle

klueska commented:
/cc @cbf123

klueska commented:
/cc @lmdaly @nolancon

cbf123 commented on pkg/kubelet/cm/devicemanager/topology_hints_test.go:
```diff
@@ -250,6 +271,65 @@ func TestGetTopologyHints(t *testing.T) {
 			},
 		},
+		{
+			description:   "Single device type, all already allocated to container",
```
Just curious, can we hit this case in practice?

klueska replied:
For the exact case you've run into -- after a kubelet restart.

klueska replied:
Ah. That makes more sense.

Yeah, that case should never come up in practice today, but it's good to test to make sure that the logic does what's expected should any code change that affects this.

cbf123 reviewed (approved):
Looks good to me, appreciate the fix.

klueska commented:
@cbf123 would you mind pulling down this branch and giving it a spin to make sure it addresses your issue in a live setting?

klueska commented:
@cbf123 after sleeping on this solution, I think there may still be a problem.

It's possible that a kubelet goes down *after* a container has been granted a set of devices, but *before* a container has been granted a set of CPUs (or vice versa). When this happens, we want the hints generated for the container after the kubelet restart to be the same as they were before the restart so that any unallocated resources (either a set of devices or CPUs in this case) can have their hints combined properly by the `TopologyManager`. Simply returning a `nil` implies a "don't care" and has the possibility to mess up alignment under this edge case.

I will rework this patch sometime today and let you know once it's done.

cbf123 commented:
Basic initial testing shows the previous version of the fix does seem to resolve the original problem.  I'll give this one a try too.

cbf123 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -320,7 +320,94 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod v1.Pod, container v1.
 		return nil
 	}
 
-	// Get a list of available CPUs.
```
do we really want to remove this comment?

cbf123 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -320,7 +320,94 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod v1.Pod, container v1.
+	//    generate hints to the TopologyManager.
+	// 2) CPU and device allocation always occurs *after* hint generation
+	//    occurs (including the case where these resources were allocated
+	//    before the kubelet was duty-cylced).
```
nit: should be duty-cycled

cbf123 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -397,3 +484,33 @@ func (p *staticPolicy) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, req
 	return hints
 }
+
+func (p *staticPolicy) isContainerMapSynced(s state.State) bool {
```
this probably deserves a comment explaining what it does and what determines the return value

cbf123 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -397,3 +484,33 @@ func (p *staticPolicy) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, req
+		return true
+	}
+
+	return false
```
any reason not to use "return len(assigned) == numFound" ?

cbf123 commented:
I'm not sure if it's because of this patch (will verify), but I'm seeing something weird and it'd be great if you could double-check.

Kubelet is running with --cpu-manager-policy=static --topology-manager-policy=single-numa-node

I started a pod that allocated 6 CPUs, and it was affined to those CPUs.  I then restarted kubelet ("sudo systemctl restart kubelet"), and the pod transitioned to "Topology Affinity Error".  Incidentally, there were many unallocated CPUs available on the same NUMA node.

cbf123 commented:
I did some additional testing.  With the previous version of the patch I was able to restart kubelet (or reboot the node) and a running pod with Guaranteed QoS and a few exclusive CPUs stayed running (or was restarted) as expected.

With the updated version of the patch, if I restart kubelet or reboot the node the pod transitions to a "Topology Affinity Error" status.

As a caveat, it's possible this is due to my own error, as I'm backporting the changes to kubernetes 1.16.2.  Will take another look at my backport job.

klueska commented:
It's a legitimate error and I could see how it would happen with the `single-numa-node` policy with the updated patch. I've been playing around with the `CPUManager` most of the day, and the deeper I dig, the more I get the feeling we aren't going to be able to deal with this properly until the following is complete: https://github.com/kubernetes/kubernetes/issues/83476

cbf123 commented:
I ran into something else with this patch and the "restricted" topology manager policy....I have 4 cpus free on NUMA node 0 and a dozen on NUMA node 1, and I tried to start up a pod asking for 6 CPUs.

Without the patch creating the pod resulted in an affinity mask of {0000000000000000000000000000000000000000000000000000000000000010 true} and allocated CPUs from NUMA 1.

With this patch creating the pod resulted in an affinity mask of {0000000000000000000000000000000000000000000000000000000000000011 true} and allocated CPUs from both NUMA 0 and NUMA 1.

klueska commented:
Just pushed a new patch. Please take a look.

I've tested it locally under a number of different scenarios and policies and it seems to be robust.

I was worried about this code https://github.com/kubernetes/kubernetes/pull/84525/files#diff-f1223d0d0c42e67041357f172a322ab6R326 not doing what was expected (because of pod statues not being up to date at this stage), but it seems to work as desired.

cbf123 commented:
Okay,  will take it for a spin.

cbf123 reviewed (commented):
seems like this new function might want some testing

cbf123 commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -239,15 +239,61 @@ type reconciledContainer struct {
+		return
+	}
+
+	activeContainers := make(map[string]struct{})
```
just wondering, why are we mapping a string to a struct here instead of something simple like a bool?

klueska replied:
It's more space efficient since i don't need the value, just the key:
https://medium.com/@l.peppoloni/how-to-improve-your-go-code-with-empty-structs-3bd0c66bc531

cbf123 commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -239,15 +239,61 @@ type reconciledContainer struct {
+		}
+	}
+	for containerID := range m.state.GetCPUAssignments() {
+		if _, ok := activeContainers[containerID]; !ok {
```
if activeContainers were mapped to a bool, I think you could just do "if activeContainers[containerID]"

klueska replied:
I had it that way to start with, but it doesn't actually work if there is an error (it panics).
It panics because when `containerID` is not present in the map, then `pod` is nil.

klueska replied:
Normally I would have fixed the bug in one commit, and then done the move to a new function in a second commit, but I forgot about this subtle change. Thanks for noting it.

cbf123 commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -41,6 +44,18 @@ func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) map[s
 				continue
 			}
 
+			allocated := m.getAllocatedDevices(string(pod.UID), container.Name, resource)
```
I think a comment about short-circuiting here would make sense.

nolancon commented:
/lgtm

lmdaly commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -41,6 +44,18 @@ func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) map[s
 			}
 
+			allocated := m.getAllocatedDevices(string(pod.UID), container.Name, resource)
+			if allocated.Len() > 0 {
```
Why in this case do we send back nil hints? Or in what case can this occur?

klueska replied:
It's not exactly `nil` hints, its an empty list of hints, which the `TopologyManager` interprets differently than `nil`. It's the way of encoding an "error" in the sense that no valid hints could be generated by this hint provider.

`nil` == I don't care about affinity, do whatever you want with this information
`[]topologymanager.TopologyHint{}` == I care about affinity, and I have no hints to give

That said, yeah, this case should ever actually come up in practice. Could even be replaced by a panic if we wanted.

lmdaly replied:
Oh okay if its interpreted as an error then that makes sense to me. Thanks for the clarification!

lmdaly commented:
Tested this out, it looks good to me.
@cbf123 if you are happy with this change I think we can go ahead and merge then?

klueska commented:
Other than maybe adding more comments as @cbf123 has suggested, there are two places I'd like some more thorough review on (hopefully from @ConnorDoyle) before merging.

1) Calling `activePods()` and `GetPodStatus()` while holding the lock in the `CPUManager`:
https://github.com/kubernetes/kubernetes/pull/84525/files#diff-67c46e7cbb41ea5e675000a3298e6b32R252

2) Whether we run the risk of not having a valid pod status in `GetTopologyHints()` such that we miss out on recreating the proper hints for container that already has CPUs allocated to it.
https://github.com/kubernetes/kubernetes/pull/84525/files#diff-f1223d0d0c42e67041357f172a322ab6R326

The reason for holding the lock in (1) is to avoid the case where a *new* pod gets added to the active list of pods (and subsequently added via AddContainer()) concurrent to this function running. If that were to happen, we would end up inadvertently removing the state for that new container, which the lock protects us from doing.

The reason for doing (2) is because we store CPU state information keyed off of `containerID` instead of (`podUID`, `containerName`)  as the devicemanager does. Once we move to a model where we allocate CPUs at pod admit time rather than at container start time, we will have to change state to be stored by (`podUID`, `containerName`) just as the devicemanager does (because there is not `containerID` information at pod admit time).  I have a WIP PR that does this already.

cbf123 commented:
I was just about to report that my testing looked decent, when I did one more test and hit a snag.
Kubelet was using the "single-numa-node" policy, I had four nodes running using exclusive CPUs.  I ran "sudo systemctl restart kubelet" and one of the pods died and now shows a status of "Topology Affinity Error".

The logs show this:
2019-11-01T17:43:19.700 controller-0 kubelet[418798]: info I1101 17:43:19.700302  418798 topology_manager.go:317] [topologymanager] Pod QoS Level: Guaranteed
2019-11-01T17:43:19.700 controller-0 kubelet[418798]: info I1101 17:43:19.700411  418798 topology_manager.go:194] [topologymanager] Hint Provider has no preference for NUMA affinity with any resource
2019-11-01T17:43:19.700 controller-0 kubelet[418798]: info I1101 17:43:19.700556  418798 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 860c4ffd-56d5-4312-ac76-72ed1fac6e7a, container appcntr1)
2019-11-01T17:43:19.700 controller-0 kubelet[418798]: info I1101 17:43:19.700660  418798 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000000000000000011 false}

klueska commented:
@cbf123 

I spent some time today trying to reproduce the error you are seeing, but was unsuccessful. The script I have been using to test is here: https://gist.github.com/klueska/7a1ff939d4b5afcd8d581de220a8db01

Can you give me some more details on how to reproduce your error? I'm not sure if this is relevant to your tests, but it's worth noting that you can't change the topology manager policy between kubelet restarts, and expect that all containers will successfully come back online. If the CPUs that were allocated to the container with a previous policy are invalid in the new policy, you would expect a failure similar to what you are seeing.

> I'm not sure where the "Hint Provider has no preference for NUMA affinity with any resource" is coming from...maybe we need to name the hint providers so these logs make more sense?

Yes, I agree, the logging in all of this should be more consistent and informative overall.

cbf123 commented:
It took me a few tries but I was able to reproduce it locally.  Note that this is with v1.16.2 and your changes backported on top.

The topology of the system is as follows:
LOGICAL CPU TOPOLOGY:
   cpu_id :   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27 
socket_id :   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   1   1   1   1   1   1   1   1   1   1   1   1   1 

Kubelet is running with the following:

--feature-gates TopologyManager=true --cpu-manager-policy=static --topology-manager-policy=single-numa-node --system-reserved=cpu=2,memory=16500Mi --kube-reserved=cpu=8

Thus, the available CPU pool for application pods is cpus 10-27

Initially I had six pods running, 5 of them using exclusive CPUs and one not.
controller-0:~$ taskset -apc 118931
pid 118931's current affinity list: 27   <---this is the non-exclusive one
controller-0:~$ taskset -apc 286290
pid 286290's current affinity list: 10-12
controller-0:~$ taskset -apc 286948
pid 286948's current affinity list: 14-19
controller-0:~$ taskset -apc 287737
pid 287737's current affinity list: 13,20-22
controller-0:~$ taskset -apc 288175
pid 288175's current affinity list: 23
controller-0:~$ taskset -apc 288553
pid 288553's current affinity list: 24-26


I restarted kubelet at system time 2019-11-04T17:44:23.  From the kubelet logs:

2019-11-04T17:44:27.366 controller-0 kubelet[324844]: info I1104 17:44:27.366105  324844 cpu_manager.go:132] [cpumanager] detected CPU topology: &{28 28 2 map[0:{0 0 0} 1:{0 0 1} 2:{0 0 2} 3:{0 0 3} 4:{0 0 4} 5:{0 0 5} 6:{0 0 6} 7:{0 0 7} 8:{0 0 8} 9:{0 0 9} 10:{0 0 10} 11:{0 0 11} 12:{0 0 12} 13:{0 0 13} 14:{1 1 14} 15:{1 1 15} 16:{1 1 16} 17:{1 1 17} 18:{1 1 18} 19:{1 1 19} 20:{1 1 20} 21:{1 1 21} 22:{1 1 22} 23:{1 1 23} 24:{1 1 24} 25:{1 1 25} 26:{1 1 26} 27:{1 1 27}]}

2019-11-04T17:44:27.366 controller-0 kubelet[324844]: info I1104 17:44:27.366363  324844 policy_static.go:120] [cpumanager] reserved 10 CPUs ("0-9") not available for exclusive assignment
2019-11-04T17:44:27.366 controller-0 kubelet[324844]: info I1104 17:44:27.366408  324844 state_mem.go:36] [cpumanager] initializing new in-memory state store
2019-11-04T17:44:27.366 controller-0 kubelet[324844]: info I1104 17:44:27.366537  324844 state_mem.go:84] [cpumanager] updated default cpuset: "27"

2019-11-04T17:44:27.366 controller-0 kubelet[324844]: info I1104 17:44:27.366549  324844 state_mem.go:92] [cpumanager] updated cpuset assignments: "map[1e1ea5076edfbb09860f1acb219265b456a4d26eedbff4c7150331e21106b5e4:13,20-22 9551426b82608372078ab3fae5b33cf895a9de0656c3ff14916cb95d48cc74f4:23 c4c6d57a0a9cf323b9a1e02174d725168e4f9032b3c85ce6ded66d5762e087de:14-19 cc51285495ec01e17ebdd856404dcebaf1649f6c3811e6f377be0dc603017071:10-12 db7531570c41a2c60883b7074faf9f13a61a0faab999ff0e4c8b5c13852859fe:24-26]"

2019-11-04T17:44:28.180 controller-0 kubelet[324844]: info I1104 17:44:28.180750  324844 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 83528366-62c6-4db2-9ad6-c5aa6782519f, container appcntr1)
2019-11-04T17:44:28.180 controller-0 kubelet[324844]: info I1104 17:44:28.180787  324844 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000000000000000001 true}
2019-11-04T17:44:28.180 controller-0 kubelet[324844]: info I1104 17:44:28.180802  324844 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 83528366-62c6-4db2-9ad6-c5aa6782519f are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000001 true}]

2019-11-04T17:44:28.181 controller-0 kubelet[324844]: info I1104 17:44:28.181285  324844 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 90c7ef5c-d6a2-46bb-9dd7-985b8605564c, container appcntr1)
2019-11-04T17:44:28.181 controller-0 kubelet[324844]: info I1104 17:44:28.181326  324844 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000000000000000010 true}
2019-11-04T17:44:28.181 controller-0 kubelet[324844]: info I1104 17:44:28.181334  324844 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 90c7ef5c-d6a2-46bb-9dd7-985b8605564c are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000010 true}]

2019-11-04T17:44:28.182 controller-0 kubelet[324844]: info I1104 17:44:28.181811  324844 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 58da3882-fa5e-43bc-9605-35646b64540c, container appcntr1)
2019-11-04T17:44:28.182 controller-0 kubelet[324844]: info I1104 17:44:28.182841  324844 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000000000000000011 false}

**Note the topology mask above, it should be "000010 true".  This pod's container had been previously affined to CPUs 13,20-22**

2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183054  324844 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 2d33431d-10ac-493e-
9a96-bf09da6c8e91, container appcntr1)
2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183134  324844 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000
000000000000010 true}
2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183147  324844 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 2d33431d-10ac-493e-9a96-bf09da6c8e91 are map[a
ppcntr1:{0000000000000000000000000000000000000000000000000000000000000010 true}]

2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183751  324844 topology_hints.go:69] [cpumanager] Regenerating TopologyHints for CPUs already allocated to (pod 029cdf5c-c6dc-4010-
a939-ec122ab69c8a, container appcntr1)
2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183858  324844 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000000000000000010 true}
2019-11-04T17:44:28.183 controller-0 kubelet[324844]: info I1104 17:44:28.183931  324844 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 029cdf5c-c6dc-4010-a939-ec122ab69c8a are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000010 true}]


At this point I assume pod 58da3882-fa5e-43bc-9605-35646b64540c went into "Topology Affinity Error" status but there's no explicit error log for it.  (There should really be one.)  The running container for that pod was killed.  All the other containers kept running as you'd expect.

One thing that may be of note is that there are a bunch of logs from reconciler.go  of the form "operationExecutor.VerifyControllerAttachedVolume started for volume..." right before the "failing" log at 2019-11-04T17:44:28.182.

klueska commented:
@cbf123 

According to your numbers:
```
cpu_id : 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27
socket_id : 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1
```

CPU 13 is on Socket 0, while CPUs 20-27

Are you sure that that container was originally started with `--topology-manager-policy=single-numa-node`?

cbf123 commented:
Hmm...good catch.  The CPUs do seem to be crossing NUMA nodes.  Here are the logs from earlier when it was created:

2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.325881  241826 topology_manager.go:308] [topologymanager] Topology Admit Handler
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.325896  241826 topology_manager.go:317] [topologymanager] Pod QoS Level: Guaranteed
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.325924  241826 topology_manager.go:194] [topologymanager] Hint Provider has no preference for NUMA affinity with any resource
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.326013  241826 topology_hints.go:80] [cpumanager] TopologyHints generated for pod 'testpod-exclusive3', container 'appcntr1': [{000
0000000000000000000000000000000000000000000000000000000000001 true} {0000000000000000000000000000000000000000000000000000000000000010 true} {0000000000000000000000000000000000000000000000000000000000000011
 false}]
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.326036  241826 topology_manager.go:285] [topologymanager] ContainerTopologyHint: {0000000000000000000000000000000000000000000000000
000000000000001 true}
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.326043  241826 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 58da3882-fa5e-43bc-9605-35646b64540c are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000001 true}]

2019-11-04T17:39:04.839 controller-0 kubelet[241826]: info I1104 17:39:04.839299  241826 policy_static.go:286] [cpumanager] Pod 58da3882-fa5e-43bc-9605-35646b64540c, Container appcntr1 Topology Affinity is: {0000000000000000000000000000000000000000000000000000000000000001 true}
2019-11-04T17:39:04.839 controller-0 kubelet[241826]: info I1104 17:39:04.839311  241826 policy_static.go:352] [cpumanager] allocateCpus: (numCPUs: 4, socket: 0000000000000000000000000000000000000000000000000000000000000001)
2019-11-04T17:39:04.839 controller-0 kubelet[241826]: info I1104 17:39:04.839472  241826 state_mem.go:84] [cpumanager] updated default cpuset: "23-27"
2019-11-04T17:39:04.840 controller-0 kubelet[241826]: info I1104 17:39:04.839895  241826 policy_static.go:385] [cpumanager] allocateCPUs: returning "13,20-22"
2019-11-04T17:39:04.840 controller-0 kubelet[241826]: info I1104 17:39:04.839909  241826 state_mem.go:76] [cpumanager] updated desired cpuset (container id: 1e1ea5076edfbb09860f1acb219265b456a4d26eedbff4c7150331e21106b5e4, cpuset: "13,20-22")
2019-11-04T17:39:04.840 controller-0 kubelet[241826]: info I1104 17:39:04.840250  241826 policy_static.go:295] [cpumanager] guaranteed: AddContainer (namespace: default, pod: testpod-exclusive3, container: appcntr1, container id: 1e1ea5076edfbb09860f1acb219265b456a4d26eedbff4c7150331e21106b5e4); numCPUS=4, cpuset=13,20-22

cbf123 commented:
So maybe the kubelet restart is fine after your patch but the initial assignment of CPUs is broken.

cbf123 commented:
I notice that staticPolicy.allocateCPUs() doesn't error out if it can't get enough NUMA-affined CPUs.  I wonder if there's a race somewhere between the "admit" check and actually allocating specific CPUs.

klueska commented:
There shouldn't be a race between containers in different pods, but there's definitely a race between containers in the same pod. That's the goal of https://github.com/kubernetes/kubernetes/issues/83476 to address.

Are all of your containers from different pods?

cbf123 commented:
I think I might see what's going on.  I started 5 pods in close succession.  It looks like the "admit" code runs for all the pods and selects a NUMA node for each:

2019-11-04T17:38:52.172 controller-0 kubelet[241826]: info I1104 17:38:52.172444  241826 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 83528366-62c6-4db2-9ad6-c5aa6782519f are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000001 true}]
2019-11-04T17:38:55.411 controller-0 kubelet[241826]: info I1104 17:38:55.411614  241826 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 90c7ef5c-d6a2-46bb-9dd7-985b8605564c are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000010 true}]
2019-11-04T17:38:58.326 controller-0 kubelet[241826]: info I1104 17:38:58.326043  241826 topology_manager.go:330] [topologymanager] Topology Affinity for Pod: 58da3882-fa5e-43bc-9605-35646b64540c are map[appcntr1:{0000000000000000000000000000000000000000000000000000000000000001 true}]

Note that at this point we have two pods assigned to NUMA node 0.  Neither of them has actually allocated any specific CPUs yet.

Now we allocate specific CPUs for pod 83528366....(the first of the two assigned to NUMA node 0) which consumes three of the four available CPUs on that NUMA node.

2019-11-04T17:38:58.630 controller-0 kubelet[241826]: info I1104 17:38:58.630171  241826 policy_static.go:352] [cpumanager] allocateCpus: (numCPUs: 3, socket: 0000000000000000000000000000000000000000000000000000000000000001)
2019-11-04T17:38:58.630 controller-0 kubelet[241826]: info I1104 17:38:58.630384  241826 state_mem.go:84] [cpumanager] updated default cpuset: "13-27"
2019-11-04T17:38:58.630 controller-0 kubelet[241826]: info I1104 17:38:58.630838  241826 policy_static.go:385] [cpumanager] allocateCPUs: returning "10-12"

A while later we try to allocate the CPUs for pod 58da3882... and we allocate the remaining CPU from NUMA node 0, then fall back to allocating the rest of the CPUs from NUMA node 1.

2019-11-04T17:39:04.839 controller-0 kubelet[241826]: info I1104 17:39:04.839311  241826 policy_static.go:352] [cpumanager] allocateCpus: (numCPUs: 4, socket: 0000000000000000000000000000000000000000000000000000000000000001)
2019-11-04T17:39:04.839 controller-0 kubelet[241826]: info I1104 17:39:04.839472  241826 state_mem.go:84] [cpumanager] updated default cpuset: "23-27"
2019-11-04T17:39:04.840 controller-0 kubelet[241826]: info I1104 17:39:04.839895  241826 policy_static.go:385] [cpumanager] allocateCPUs: returning "13,20-22"

cbf123 commented:
So if I had to guess, I'd say that *this* patch series is probably fine, but maybe I should open up another issue covering the race condition during creation of multiple pods in close succession.

cbf123 commented:
And yeah, these are all separate pods.

klueska commented:
https://github.com/kubernetes/kubernetes/issues/83476 should address this race between pods as well.

Unfortunately, I don't think there is any way out of this race for 1.16, as it boils down to the fact that Hints are generated at pod admit time, while allocation happens at container start time.

I guess the best we could o is error out at container creation if the CPUs being allocated don't actually align with what was expected from the Hint. At least then you error out without allocation CPUs that don't have the alignment you expected.

klueska commented:
If you're satisfied with this patch for now at least, can you give it an lgtm so other reviewers know the state of things.

cbf123 reviewed (approved):
LGTM

klueska commented:
> I guess the best we could do is error out at container creation if the CPUs being allocated don't actually align with what was expected from the Hint. At least then you error out without allocating CPUs that don't have the alignment you expected.

We can't actually even do that without intertwining the topologymanager policy with the allocation policy (since for the best-effort policy, for example, we want to continue the allocation even if the NUMA affinity isn't satisfied).

klueska committed:
Sync all CPU and device state before generating TopologyHints for them

This ensures that we have the most up-to-date state when generating
topology hints for a container. Without this, it's possible that some
resources will be seen as allocated, when they are actually free.

klueska committed:
Add some more comments to GetTopologyHints() in the devicemanager

klueska committed:
Ensure CPUManager TopologyHints are regenerated after kubelet restart

This patch also includes test to make sure the newly added logic works
as expected.

klueska committed:
Ensure devicemanager TopologyHints are regenerated after kubelet restart

This patch also includes test to make sure the newly added logic works
as expected.

klueska commented:
To reiterate, I'm looking for a review about the points in the following comment from someone more knowledgable in this area: https://github.com/kubernetes/kubernetes/pull/84525#issuecomment-548852523

/cc @derekwaynecarr 
/cc @ConnorDoyle

ConnorDoyle commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -66,12 +89,15 @@ func (m *ManagerImpl) deviceHasTopologyAlignment(resource string) bool {
 
+func (m *ManagerImpl) getAllocatedDevices(podUID, containerName, resource string) sets.String {
+	// Pull the list of device IDs directly from the list of podDevices.
+	return m.podDevices[podUID][containerName][resource].deviceIds
```
It's not expected for the 2nd and 3rd level mappings to be nil, but it would be nicer to check (and log and return an empty set I suppose)

klueska replied:
One better -- there was already a function for this in the `podDevices` abstraction:
```
m.podDevices.containerDevices()
```

ConnorDoyle commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -28,26 +28,49 @@ import (
 func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) map[string][]topologymanager.TopologyHint {
-	deviceHints := make(map[string][]topologymanager.TopologyHint)
+	// Garbage collect any stranded device resources before providing TopologyHints
+	m.updateAllocatedDevices(m.activePods())
```
:+1: for moving this out of `getAvailableDevices`.

ConnorDoyle reviewed (commented):
These changes make sense based on the PR description of the fault. Thanks @klueska for again leaving great comments to follow and @cbf123 for adding your review.

/approve

ConnorDoyle commented:
Responding to the specific questions called out by @klueska:

1. I don't think there's a problem with calling `activePods` or `GetPodStatus` under the cpumanager mutex. The issue with holding the lock in `removeStaleState` may be that we do perform some file I/O within that function by virtue of calling `m.policy.RemoveContainer`. Ideally we could find a way to avoid that.
1. Could you walk through the sequence of events that causes the pod status to be missing when GetTopologyHints is called on the cpumanager? As you pointed out in the comment, it seems like avoiding that possibility would have to wait until the cpumanager state is refactored to store (pod.Name, container.Name) composite keys instead of container ID. So even if it is possible, I'm not sure the problem is addressable in this patch.

/cc @derekwaynecarr for a 2nd opinion, he may have more background on pod status guarantees

klueska commented:
@ConnorDoyle Thanks for the review, and thanks for the response to my questions!

I agree theres nothing more we can do in this PR if a pod status is not ready at the point we need it to be. It would be good to know if it is at all possible to *not* have a vaild status at this point though. Let's wait and what @derekwaynecarr's response is.

With that said, I think all that's missing now is an /lgtm now that I've addressed: https://github.com/kubernetes/kubernetes/pull/84525#pullrequestreview-312201285

klueska commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

klueska commented:
/test pull-kubernetes-integration

klueska commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ConnorDoyle commented:
/lgtm

tedyu commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -239,15 +241,68 @@ type reconciledContainer struct {
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			containerID, err := findContainerIDByName(&pstatus, container.Name)
+			if err != nil {
+				// If even one container does not have it's containerID set, skip state removal.
```
I wonder why state removal cannot be done for the other containers.

klueska replied:
I'm not sure I understand the concern. This function removes state from *all* containers that don't make it into the `activeContainers` list. If there's a possibility that we might not add a container to this list when we should, then the only thing we can do is exit early and try again another time.

tedyu replied:
I think error from getting containerID for one of InitContainers shouldn't skip all of Containers

klueska replied:
I'm happy to consider a modification if you can suggest a way to do that. I don't immediately see a way to do it though.

tedyu replied:
How about the following ?
```
		for _, container := range pod.Spec.InitContainers {
			containerID, err := findContainerIDByName(&pstatus, container.Name)
			if err != nil {
				break
			}
			activeContainers[containerID] = struct{}{}
		}
		for _, container := range pod.Spec.Containers {
			containerID, err := findContainerIDByName(&pstatus, container.Name)
			if err != nil {
				// If even one container does not have it's containerID set, skip state removal.
				return
			}
			activeContainers[containerID] = struct{}{}
		}
```
If you think the above is better, I can send out a PR after this PR merges.

klueska replied:
That won't work though.

If any call to `findContainerIDByName()` fails then the `containerID` of that container won't be added to the `activeContainers` list (even though there might actually be valid state for that container).

At the end of the function we then remove any state *not* part of the `activeContainers` list, which is unacceptable if it was just a sporadic syncing of the pod state.

In any case, this entire problem goes away in the next PR I have open (there will be no need to call out to `findContainerIDByName()` going forward):
https://github.com/kubernetes/kubernetes/pull/84462

tedyu replied:
Glad you have plan for improving this part of code.

klueska commented:
/retest

klueska commented:
/retest
